File: javascript/pagination-helper.js
// Pagination Helper - 5 kyu
// TODO: complete this object/class

// The constructor takes in an array of items and a integer indicating how many
// items fit within a single page
class PaginationHelper {
	constructor(collection, itemsPerPage) {
		this.collection = collection
		this.itemsPerPage = itemsPerPage
	}

	// returns the number of items within the entire collection
	itemCount() {
		this.length = this.collection.length
		return this.length
		//return (this.collection).length
	}

	// returns the number of pages
	// divides total number of items by items per page, then rounds up to account for any remainder
	pageCount() {
		this.pages = Math.ceil(this.itemCount() / this.itemsPerPage)
		return this.pages
	}

	// returns the number of items on the current page. page_index is zero based.
	// this method should return -1 for pageIndex values that are out of range
	pageItemCount(pageIndex) {
		// if the index is higher than the page count
		this.lastPageIndex = this.pageCount() - 1
		if (pageIndex > this.lastPageIndex) {
			return -1
		}
		// if it's the last page
		else if (pageIndex === this.lastPageIndex) {
			if (this.itemCount % this.itemsPerPage === 0) {
				return this.itemsPerPage
			}
			else {
				return this.itemCount() % this.itemsPerPage
			}
		}
		// if it's any page, up to but not including the last page
		else if (pageIndex >= 0) {
			return this.itemsPerPage
		}
		
	}
	// determines what page an item is on. Zero based indexes
	// this method should return -1 for itemIndex values that are out of range
	pageIndex(itemIndex) {
		this.lastItemIndex = this.itemCount() - 1
		// if it's the first item
		if (itemIndex === 0 && this.pageCount() > 0) {
			return 0
		}
		// if index is too high
		else if (itemIndex > this.lastItemIndex || itemIndex < 0) {
			console.log(`${itemIndex} is invalid`)
			return -1
		}
		// if it's the last item
		else if (itemIndex === this.lastItemIndex) {
			console.log(`${itemIndex} is on the last page`)
			return this.pageCount()-1
		}
		// checking every item between first and last
		else if (itemIndex > -1) {
			for (let i = this.pageCount()-1; i >= 0; i--) {
			// we check for first page that is greater than itemIndex - page AFTER where item would be indexed
				if (itemIndex > (this.itemsPerPage * i)) {
					console.log(`item with index ${itemIndex} is on page ${i}`)
					console.log(i)
					return i
				}
			}
		}
	}
}


var helper = new PaginationHelper(['a', 'b', 'c', 'd', 'e', 'f'], 4);
//console.log(helper.pageCount())
//console.log(helper.itemCount())
//console.log(helper.pageItemCount(0))
//console.log(helper.pageItemCount(1))
//console.log(helper.pageItemCount(2))
// console.log(helper.pageIndex(5))
// console.log(helper.pageIndex(2))
// console.log(helper.pageIndex(0))
//console.log(helper.pageIndex(20))
//console.log(helper.pageIndex(-10))
// console.log(helper.pageIndex(4))

let list2 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24]
var helper2 = new PaginationHelper(list2, 10)
console.log(helper2.pageIndex(22))

File: python/pagination-helper.py
import math	
class PaginationHelper:
	
 	# The constructor takes in an array of items and a integer indicating
  	# how many items fit within a single page
	def __init__(self, collection, items_per_page):
		self.collection = collection 
		self.items_per_page = items_per_page
	  
   	# returns the number of items within the entire collection
	def item_count(self):
		return len(self.collection)
	  
   	
	# returns the number of pages
	def page_count(self):
		return math.ceil(self.item_count()/self.items_per_page)

	# returns the number of items on the current page. page_index is zero based
	# this method should return -1 for page_index values that are out of range
	def page_item_count(self, page_index):
		if page_index > (self.page_count()-1) or page_index < 0:
			return -1
		if self.item_count()%self.items_per_page == 0:
			return self.items_per_page
		elif self.item_count()%self.items_per_page != 0:
			if page_index == (self.page_count()-1):
				return self.item_count()%self.items_per_page
			else:
				return self.items_per_page
	   
  
	# determines what page an item is on. Zero based indexes.
	# this method should return -1 for item_index values that are out of range
	def page_index(self, item_index):
		if item_index > (self.item_count()-1) or item_index < 0:
			return -1
		if item_index == (self.item_count()-1):
			return self.page_count()-1
		else:
			return int(item_index/self.items_per_page)
  

# helper = PaginationHelper(['a','b','c','d','e','f'], 4)
# print(helper.page_count()) # 2 pages. 0 indexed, page 0, page 1
# print(helper.item_count()) # 
# print(helper.page_item_count(2)) # -1
# print(helper.page_item_count(1)) # 2
# print(helper.page_item_count(0)) # 4

# print(helper.page_index(20)) # -1
# print(helper.page_index(5))  # should == 1 (zero based index)
# print(helper.page_index(2))  # should == 0
# print(helper.page_index(-10))  # should == -1 because negative indexes are invalid

